lib/auth/touchid/api.go

```diff
@@ -0,0 +1,365 @@
+	}
+	ok := false
+	for _, param := range cc.Response.Parameters {
+		// ES256 is all we can do.
+		if param.Type == protocol.PublicKeyCredentialType && param.Algorithm == webauthncose.AlgES256 {
+			ok = true
+		}
+	}
+	if !ok {
+		return nil, errors.New("cannot fulfil credential parameters, only ES256 are supported")
+	}
+
+	rpID := cc.Response.RelyingParty.ID
+	user := cc.Response.User.Name
+	userHandle := cc.Response.User.ID
+
+	// TODO(codingllama): Handle double registrations and failures after key
+	//  creation.
+	resp, err := native.Register(rpID, user, userHandle)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	credentialID := resp.CredentialID
+	pubKeyRaw := resp.publicKeyRaw
+
+	// Parse public key and transform to the required CBOR object.
+	pubKey := pubKeyFromRawAppleKey(pubKeyRaw)
+	x := make([]byte, 32) // x and y must have exactly 32 bytes in EC2PublicKeyData.
+	y := make([]byte, 32)
+	pubKey.X.FillBytes(x)
+	pubKey.Y.FillBytes(y)
+
+	pubKeyCBOR, err := cbor.Marshal(
+		&webauthncose.EC2PublicKeyData{
+			PublicKeyData: webauthncose.PublicKeyData{
+				KeyType:   int64(webauthncose.EllipticKey),
+				Algorithm: int64(webauthncose.AlgES256),
+			},
+			// See https://datatracker.ietf.org/doc/html/rfc8152#section-13.1.
+			Curve:  1, // P-256
+			XCoord: x,
+			YCoord: y,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	attData, err := makeAttestationData(
+		protocol.CreateCeremony, origin, rpID, cc.Response.Challenge,
+		&credentialData{
+			id:         credentialID,
+			pubKeyCBOR: pubKeyCBOR,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sig, err := native.Authenticate(credentialID, attData.digest)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	attObj, err := cbor.Marshal(protocol.AttestationObject{
+		RawAuthData: attData.rawAuthData,
+		Format:      "packed",
+		AttStatement: map[string]interface{}{
+			"alg": int64(webauthncose.AlgES256),
+			"sig": sig,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &wanlib.CredentialCreationResponse{
+		PublicKeyCredential: wanlib.PublicKeyCredential{
+			Credential: wanlib.Credential{
+				ID:   credentialID,
+				Type: string(protocol.PublicKeyCredentialType),
+			},
+			RawID: []byte(credentialID),
+		},
+		AttestationResponse: wanlib.AuthenticatorAttestationResponse{
+			AuthenticatorResponse: wanlib.AuthenticatorResponse{
+				ClientDataJSON: attData.ccdJSON,
+			},
+			AttestationObject: attObj,
+		},
+	}, nil
+}
+
+func pubKeyFromRawAppleKey(pubKeyRaw []byte) *ecdsa.PublicKey {
+	// "For an elliptic curve public key, the format follows the ANSI X9.63
+	// standard using a byte string of 04 || X || Y. (...) All of these
+	// representations use constant size integers, including leading zeros as
+	// needed."
+	// https://developer.apple.com/documentation/security/1643698-seckeycopyexternalrepresentation?language=objc
+	pubKeyRaw = pubKeyRaw[1:] // skip 0x04
```

Comment:
Do we validate that `pubKeyRaw` is non-empty at some point prior to calling this? Just thinking we should prevent a panic if for some reason we get an empty/nil slice.